### PR TITLE
Allow trailing coma in asserts

### PR DESCRIPTION
**Pull request by deadalnix**
Because

### CODE REVIEW

**Comment by Hackerpilot** [comment]
This is both ugly and consistent with the way that function argument/parameter parsing works...

**Comment by yebblies** [comment]
I guess this is next?
```d
if (x,)
{
}
```

**Comment by deadalnix** [comment]
No

**Comment by WalterBright** [comment]
> Because

Need a more compelling justification.

**Comment by deadalnix** [comment]
@WalterBright Simply the ability to have consistent formatting when passing a series of argument to something.

**Comment by AndrejMitrovic** [comment]
It's a little early for april fools!

> @WalterBright Simply the ability to have consistent formatting when passing a series of argument to something.

If you're talking about mixins you can use `array.join(",");`, though you might have to change a foreach loop into something simpler.

**Comment by deadalnix** [comment]
I'm not talking about mixin. If I was talking about mixin I would have used the word mixin. I'm talking about being able to format argument list consistently. This has been explained both by myself and @Hackerpilot , if you have something to say about this, please do, but if you plan to poison the well by using dismiss terms like "april fool" make sure to do it in a way one doesn't have to question your ability to pass the Turing test.

**Comment by AndrejMitrovic** [comment]
Take it easy..

Anyway if it's consistent with everything else DMD does then maybe it should be allowed indeed.

**Comment by deadalnix** [comment]
No hard feelings.

**Comment by Hackerpilot** [comment]
> @Hackerpilot , if you have something to say about this, please do, but if you plan to poison the well by using dismiss terms like "april fool" make sure to do it in a way one doesn't have to question your ability to pass the Turing test.

Wrong person.

**Comment by deadalnix** [comment]
There is a comma.

> This has been explained both by myself and @Hackerpilot , ...

**Comment by deadalnix** [comment]
Trailing comma are useful. I do not really wish to go back on this argument, it has been debated to death already in all languages and file format that exists. There are good usability reasons and blame conservation reasons.

D already has trailing comma and it is for the best. This is about removing inconsistency.

**Comment by yebblies** [comment]
> Trailing comma are useful. I do not really wish to go back on this argument, it has been debated to death already in all languages and file format that exists. There are good usability reasons and blame conservation reasons.

I understand why a trailing comma is useful in enum declarations, function argument lists, array literals etc.  But assert is not a function, and does not have variadic arguments.

My though is that this is good style:
```d
int[] x =
[
    1,
    2,
    3,
    4,
];
```
but this is not:
```d
assert(x, "message",);
```
I'm expecting the first would be preferred, and the latter would be rejected during code review.

If I'm not the only one that thinks this, then it would seem we're adding support for a syntax that should never be used.

So, do any of the "good usability reasons and blame conservation reasons" actually apply to assert?

**Comment by deadalnix** [comment]
@yebblies , both at work and in various projects of mine, I use trailing comma for multiple lines argument lists, and no trailing comma for single line ones.

    foo(a, b,) // bad style
    foo(
      a,
      b,
    ) // good style.

It is relevent when a and b are actually long expressions rather than single letters.

While assert is not a function it is to be noted we do this for all argument lists in D, not only function's. We do it for array lterals, for template arguments, for templateand function parameters, and more. Assert is the exception. I do not wish to special case assert formating when condition/message are long.

More generally, I think it is not good to allow or disallow trailing comma, or prety much anything else on a case by case basis. That makes the language more complex for no good benefit.

**Comment by AndrejMitrovic** [comment]
> No hard feelings.

FYI I've always admired you @deadalnix and it really felt like your comment came out of left-field. I'm sorry if my own comment made it look like I had bad intentions.

I can understand the frustration with making DMD pulls which leads to one to being pessimistic (been there, done that..), so let's just say fuck it and bury the hatchets?

I agree that we really need to keep these rules consistent. The more special cases involved the more things we need to keep in our head.

Btw @deadalnix, is there any way we can make this implementation generic instead of specializing for `TOKassert`? Question goes to @9rnsr too.

**Comment by deadalnix** [comment]
@AndrejMitrovic Yes, let's burry the hatchet. Sorry for being hot blooded.

As to make this more generic, I don't know much what it involve. I'm only somewhat familiar with DMD source code. if someone more experienced with DMD suggest a way to make this more generic, I'd be happy to do it.

**Comment by AndrejMitrovic** [comment]
> @AndrejMitrovic Yes, let's burry the hatchet. Sorry for being hot blooded.

Cool. :)

>  if someone more experienced with DMD suggest a way to make this more generic, I'd be happy to do it.

Don't worry about it too much, we can live with the special case for now (obviously it was never generic to begin with). Just make it go green so we can merge it.

Btw, anyone know what Kenji is up to? Apparently he hasn't had any activity in over 6 months. Hope he's just busy with other things now. :)

**Comment by deadalnix** [comment]
Updated, that should be green soon.

**Comment by deadalnix** [comment]
Ping ?

**Comment by Hackerpilot** [comment]
To clarify my position: I don't like trailing commas in argument lists, but that doesn't matter because we already have them. I am in favor of this change because consistency is important.

File: src/parse.d [hunk 1]
```diff
@@ -1920,7 +1920,11 @@ public:
         if (token.value == TOKcomma)
         {
             nextToken();
-            msg = parseAssignExp();
+            if (token.value != TOKrparen)
+            {
+                msg = parseAssignExp();
+                if (token.value == TOKcomma) nextToken();
```
Comment (9rnsr) [inline src/parse.d]: In dmd code style, if-condition and its body are placed in their own lines.

File: src/parse.d [hunk 2]
```diff
@@ -6902,7 +6906,11 @@ public:
                 if (token.value == TOKcomma)
                 {
                     nextToken();
-                    msg = parseAssignExp();
+                    if (token.value != TOKrparen)
+                    {
+                        msg = parseAssignExp();
+                        if (token.value == TOKcomma) nextToken();
```
Comment (9rnsr) [inline src/parse.d]: Ditto.

File: test/compilable/compile1.d [hunk 3]
```diff
@@ -939,3 +939,14 @@ struct S3438_2 { this(int x, ...) { } }
...
+
+/***************************************************/
+// 15362
+
+void func15362()
+{
+    assert(true);
+    assert(true,);
+    assert(true, "So true");
+    assert(true, "Very, very true",);
```
Comment (9rnsr) [inline test/compilable/compile1.d]: `static assert` needs to be tested, too.

**Comment by 9rnsr** [comment]
I've come back.

> is there any way we can make this implementation generic instead of specializing for TOKassert? 

Currently `parseParameters` wouldn't work well against the fixed-number arguments. So making implementation more generic would need more work. In other words, this PR is enough simple as the new behavior implementation.

LGTM.

**Comment by deadalnix** [comment]
Done, welcome back @9rnsr !

**Comment by 9rnsr** [comment]
LGTM. I'm on a positive stance to the change. When an `assert` code is generated by string mixin, reducing special cases for a trailing comma would be useful.

**Comment by AndrejMitrovic** [comment]
Auto-merge toggled on

**Comment by deadalnix** [comment]
<3

**Pull request merged**